[BEAM-3304] Go triggering support

riteshghorse opened the pull request:
Trigger support for Go SDK

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon">
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/badge/icon?subject=V1+Streaming">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon?subject=V1+Java+11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/badge/icon?subject=V2+Streaming">
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon?subject=Java+8">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon?subject=Java+11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon?subject=Portable+Streaming">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/badge/icon?subject=Portable">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon?subject=Structured+Streaming">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon">
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon?subject=ValCont">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon?subject=Portable">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/badge/icon?subject=V1">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/badge/icon?subject=V1+Java11">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/badge/icon?subject=V2">
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon?subject=3.6">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon?subject=3.7">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon?subject=3.8">
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon">
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon?subject=Tests">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon?subject=Lint">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon?subject=Docker">
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon?subject=Docs">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          <img alt="Build Status" src="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/badge/icon">
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

riteshghorse commented:
R: @lostluck

lostluck commented on sdks/go/pkg/beam/core/graph/window/strategy.go:
```diff
@@ -16,12 +16,16 @@
 	Fn *Fn
-
 	// TODO(BEAM-3304): trigger support
+	Trigger *pipepb.Trigger
```
As a rule, we avoid having users have access to, or deal with, the proto directly, including via any of the built up structs. The protos are an implementation detail, not an interface for users to use.

This largely means building up SDK specific mechanisms that are then translated to the proto in graphx/translate.go. Essentially, the graph/** directories shouldn't be depending on the protos.

That said, this is fine during development as Pane plumbing and such is being developed.

lostluck reviewed (commented):
I think you're missing a few files, or it's out of date? This a good start for validating and getting a test working though. As written I can't see how you'd get to a pane encoding error however, hence guessing files are missing.

Feel free to add a new test and pipeline functions to the windowinto.go files, in order to test out triggers. It will certainly fail on the Go direct runner, so also add a filter to prevent it from running on the direct runner during the integration tests here:
https://github.com/apache/beam/blob/5361d17cbce5e666dddd13872297e58ecb7313e1/sdks/go/test/integration/integration.go#L60

riteshghorse commented:
I changed parts in WindowMarshallingStrategy in translate.go to try out triggers. But did undo all before committing. I'll write a separate test doing this.

lostluck commented on sdks/go/test/integration/primitives/windowinto.go:
```diff
@@ -91,3 +91,35 @@ func WindowSums_GBK(s beam.Scope) {
+
+		// change below statement to: windowed.WindowingStrategy().SetAlways()
+		// to get the decoding error.
+		windowed.WindowingStrategy().SetDefault()
```
OK. It pains me to suggest this, but this user experience leaves a little to be desired and we have a short window before the Go SDK exits experimental where we can make it better. 

I propose we make a `beam.WindowIntoOption` interface type, and change `beam.WindowInto` to take a variadic parameter of those to configure the trigger and other windowing strategy properties. Then we can type assert and apply the options to the windowing strategy without much concern.

This would be similar to how `beam.ParDo` has an option type https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/option.go#L25
https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/pardo.go#L356

My recommendation would be to make that into a separate change by itself, not defining any of the concrete `beam.WindowIntoOption` types and handling, and just changing the signature to take in `...options beam.WindowIntoOption` as the final parameter (variadic parameters must be the final one in a function signature).

While this change would not be 100% backwards compatible, it would be source compatible for all reasonable uses of `beam.WindowInto`. The main place it would break is if a user were doing something like:

`var myWindowInto func(s Scope, ws *window.Fn, col PCollection) PCollection = beam.WindowInto`

which is unlikely and weird to begin with (it wouldn't serve anything beyond ensuring the signature never changes).

Then we can rebase and get back to here and finish handling the implementation.

lostluck commented on sdks/go/pkg/beam/core/graph/edge.go:
```diff
@@ -527,8 +527,8 @@ func NewImpulse(g *Graph, s *Scope, value []byte) *MultiEdge {
 // NewWindowInto inserts a new WindowInto edge into the graph.
-func NewWindowInto(g *Graph, s *Scope, wfn *window.Fn, in *Node) *MultiEdge {
-	n := g.NewNode(in.Type(), &window.WindowingStrategy{Fn: wfn}, in.Bounded())
+func NewWindowInto(g *Graph, s *Scope, wfn *window.Fn, tr window.TriggerType, in *Node) *MultiEdge {
```
NewWindowInto should just change to take in a `*window.WindowingStrategy` instead of adding more and more parameters to it. There are other modes to handle in the future and by populating the windowing strategy in the Beam package before it gets here, will save time and effort when adding features later.

riteshghorse replied:
Make sense.

lostluck commented on sdks/go/pkg/beam/windowing.go:
```diff
@@ -21,21 +21,47 @@ import (
 )
 
+type WindowIntoOption interface {
+	private()
```
"private" isn't a good name for this blocking function, because beam.Option is already using it. You can test this out yourself by seeing that you can pass `WindowIntoOption`s as an Option into beam.ParDo.

A better unexported method name would be `windowIntoOption()` to avoid this conflation.

lostluck commented on sdks/go/pkg/beam/windowing.go:
```diff
@@ -21,21 +21,47 @@ import (
 		return PCollection{}, errors.New("invalid input pcollection")
 	}
+	var edge *graph.MultiEdge
+	for _, opt := range opts {
```
Consider just creating a windowing strategy before iterating over the ops with all the reasonable defaults set, and then update that strategy with the whatever is in options. Then we don't end up with a combinatorial explosion when other options are added or with the awkward "if there are no options, do this" length check written below.

lostluck commented on sdks/go/pkg/beam/windowing.go:
```diff
@@ -21,21 +21,47 @@ import (
+
+func (t WindowTrigger) windowIntoOption() {}
+
+func (t WindowTrigger) GetName() window.TriggerType {
```
There's no need for a GetName function. As a rule, don't add "Get" to a method that just returns a field.
See https://golang.org/doc/effective_go#Getters

lostluck commented on sdks/go/pkg/beam/windowing.go:
```diff
@@ -21,21 +21,47 @@ import (
+	for _, opt := range opts {
+		switch opt.(type) {
+		case WindowTrigger:
+			edge = graph.NewWindowInto(s.real, s.scope, &window.WindowingStrategy{Fn: ws, Trigger: opt.(WindowTrigger).GetName()}, col.n)
```
Consider accessing the trigger kind directly with `.Name` without a method instead, and remove the method.

lostluck commented on sdks/go/pkg/beam/core/graph/window/trigger.go:
```diff
@@ -0,0 +1,41 @@
+
+package window
+
+type TriggerType string
```
Consider how we'd implement the more complicated triggers that have sub triggers. For example, AfterAll and AfterAny are missing their sub trigger configurations.

lostluck commented on sdks/go/pkg/beam/core/graph/window/trigger.go:
```diff
@@ -0,0 +1,41 @@
+	Default  TriggerType = "Trigger_Default_"
+	Always   TriggerType = "Trigger_Always_"
+	AfterAny TriggerType = "Trigger_AfterAny_"
+	AfterAll TriggerType = "Trigger_AfterAny_"
```
Typo: duplicated as AfterAny instead of AfterAll.

lostluck reviewed (commented):
Slowly but surely this is shaping up! I'll stop peaking in at this point and wait for you to ask me to Please Take Another Look (PTAL) before my next pass.

riteshghorse commented:
PTAL.

Added pane helpers and changed all affecting function signatures/receivers. The error of `source decode failed  cause by ____` seems to have gone. I tested this the old way (TestTriggerWindowSums_GBK). Next step would be writing isolated trigger tests.

lostluck commented on sdks/go/pkg/beam/core/graph/coder/panes.go:
```diff
@@ -0,0 +1,143 @@
+)
+
+func chooseEncoding(v typex.PaneInfo) int {
+	if (v.Index == 0 || v.NonSpeculativeIndex == 0) || v.Timing == typex.UNKNOWN {
+		return FIRST
+	} else if v.Index == v.NonSpeculativeIndex || v.Timing == typex.EARLY {
+		return ONE_INDEX
+	} else {
+		return TWO_INDEX
+	}
```
Prefer using a switch statment for things like this instead of relying on if-else ladders.
```suggestion
	switch {
	case v.Index == 0 || v.NonSpeculativeIndex == 0) || v.Timing == typex.UNKNOWN:
 		return FIRST
	case v.Index == v.NonSpeculativeIndex || v.Timing == typex.EARLY:
		return ONE_INDEX
	default:
		return TWO_INDEX
	}
```

lostluck replied:
As a result, we can probably just move these conditions into the encoder, rather than relating them through constants, and have a comment that explains their differences, which you're using as a the constant name.

lostluck commented on sdks/go/pkg/beam/core/typex/special.go:
```diff
@@ -64,6 +65,23 @@ type Window interface {
 	Equals(o Window) bool
 }
 
+type Timing string
```
Instead of using a string, consider using an `int` (or even a uint8) as the base type. It's more compact in memory. It also lets one avoid doing string comparisons to know values for encodings, and avoids some awkward conversions that this PR is currently making.

See https://yourbasic.org/golang/iota/ for a good rundown for how to do "enums" in Go. As well as in Effective Go. https://golang.org/doc/effective_go#constants

riteshghorse replied:
Thanks for this! This is really helpful.

lostluck commented on sdks/go/pkg/beam/core/typex/special.go:
```diff
@@ -64,6 +65,23 @@ type Window interface {
+type Timing string
+
+const (
+	EARLY   Timing = "EARLY"
+	ON_TIME Timing = "ON_TIME"
+	LATE    Timing = "LATE"
+	UNKNOWN Timing = "UNKNOWN"
```
Despite these being constants, it's not idiomatic go to use All caps for constants. 

I also recommend prefixing what these are related to eg. PaneEarly, PaneOnTime, PaneLate, PaneUnknown 

Mostly so they're adjacent in Go Doc. vs suffixing, which is a bit more readable.

We can only get away without prefixing because we don't have a "pane" package which would delare what they're for earlier. In this case we shouldn't have a pane package, it would be too small for practical uses.

This is because the typex package has a pretty grab bag set of values, so the enumerations need to be made clearer for their use elsewhere.

riteshghorse replied:
Got it.

lostluck commented on sdks/go/pkg/beam/core/typex/special.go:
```diff
@@ -64,6 +65,23 @@ type Window interface {
+
+type PaneInfo struct {
+	Timing              Timing
+	IsFirst             bool
+	IsLast              bool
+	Index               int64
+	NonSpeculativeIndex int64
```
Note, there's nothing wrong with the current approach, I'm pointing out a thing about Go:

In this case, you can write these fields more compactly as: 
```suggestion
	IsFirst, IsLast				bool
	Index, NonSpeculativeIndex		int64
```

lostluck commented on sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:
```diff
@@ -117,7 +118,7 @@ func (x *translator) translateTransform(trunk string, id string) ([]*df.Step, er
 		// a nested context at runtime.
 		var buf bytes.Buffer
-		if err := exec.EncodeWindowedValueHeader(exec.MakeWindowEncoder(coder.NewGlobalWindow()), window.SingleGlobalWindow, mtime.ZeroTimestamp, &buf); err != nil {
+		if err := exec.EncodeWindowedValueHeader(exec.MakeWindowEncoder(coder.NewGlobalWindow()), window.SingleGlobalWindow, mtime.ZeroTimestamp, typex.PaneInfo{}, &buf); err != nil {
```
Question: Is the zero value for your pane info type identical to the "No firing" pane we've been writing? If so, very nice!

Optionally, also add a function to the typex package: typex.NoFiringPane() to document the intended semantic meaning of the type at the call sights like this.

lostluck commented on sdks/go/pkg/beam/core/graph/window/trigger.go:
```diff
@@ -0,0 +1,42 @@
+type TriggerType string
+
+const (
+	Default  TriggerType = "Trigger_Default_"
+	Always   TriggerType = "Trigger_Always_"
+	AfterAny TriggerType = "Trigger_AfterAny_"
+	AfterAll TriggerType = "Trigger_AfterAll_"
```
WRT to the const names here, we probably want to call these DefaultTrigger, AlwaysTrigger, etc, since users will see these as  "window.Default" and "window.Always" and it won't necessarily be clear what they mean.

lostluck commented on sdks/go/pkg/beam/core/graph/window/trigger.go:
```diff
@@ -0,0 +1,42 @@
+	AfterAll TriggerType = "Trigger_AfterAll_"
+)
+
+func (ws *WindowingStrategy) SetAfterAll() {
+	ws.Trigger = AfterAll
+}
+
+func (ws *WindowingStrategy) SetAfterAny() {
+	ws.Trigger = AfterAny
+}
+
+func (ws *WindowingStrategy) SetAlways() {
+	ws.Trigger = Always
+}
+
+func (ws *WindowingStrategy) SetDefault() {
+	ws.Trigger = Default
+}
```
We can probably remove these helper methods since they aren't being called, in favour of directly setting the fields ourselves.

lostluck commented on sdks/go/pkg/beam/core/runtime/exec/coder.go:
```diff
@@ -1079,7 +1112,7 @@ func (*intervalWindowDecoder) DecodeSingle(r io.Reader) (typex.Window, error) {
 var paneNoFiring = []byte{0xf}
```
If we aren't using this variable anywhere anymore we can remove it.

lostluck commented on sdks/go/pkg/beam/core/runtime/exec/coder.go:
```diff
@@ -1088,25 +1121,22 @@ func EncodeWindowedValueHeader(enc WindowEncoder, ws []typex.Window, t typex.Eve
+func DecodeWindowedValueHeader(dec WindowDecoder, r io.Reader) ([]typex.Window, typex.EventTime, typex.PaneInfo, error) {
 	// Encoding: Timestamp, Window, Pane (header) + Element
-
+	pn := typex.PaneInfo{}
```
It would be more idiomatic to declare the zero value in place for the returns, which makes it explicit that they're the zero value. 
The reason for this is to limit scope, and make it easier on readers. If possible we keep that as small as possible. With a pre-declaration like this, a reader has to keep an eye out for it, for when it's used, and what might have affected it before it's returned. By writing our intent (return a zero value on error) directly where it's happening, we avoid this.

That said, I understand the repetition can be tiresome. You can instead have defined in this function scope helper function

`onError := func(err error) { return  nil, mtime.ZeroTimestamp, typex.PaneInfo{}, err }`

Which you can then call as `return onError(err)` and it will compile down to the same thing.

Non-pointer zeros being inconvenient is a hot topic of discussion for Go and eventually someone will propose a satisfying solution for it.

riteshghorse replied:
This is great!

lostluck commented on sdks/go/pkg/beam/core/runtime/exec/datasource.go:
```diff
@@ -134,7 +134,7 @@ func (n *DataSource) Process(ctx context.Context) error {
 		pe.Timestamp = t
 		pe.Windows = ws
-
+		pe.Pane = pn
```
Please keep the blank line before the next block. 
Blank lines used carefully help readability.

lostluck commented on sdks/go/pkg/beam/windowing.go:
```diff
@@ -21,21 +21,40 @@ import (
+		switch opt.(type) {
+		case WindowTrigger:
+			ws.Trigger = opt.(WindowTrigger).Name
+		default:
+			ws.Trigger = window.Default
```
Since this switch is only run if we have options, it's valid to return an error or even `panic`if we don't know what type of option it is. That way users won't accidently override earlier set triggers because an implementation was half done somehow.

lostluck commented on sdks/go/test/integration/primitives/windowinto.go:
```diff
@@ -91,3 +91,32 @@ func WindowSums_GBK(s beam.Scope) {
+	}
+
+	// Use fixed windows to divide the data into 3 chunks.
+	validate(s.Scope("Fixed"), window.NewFixedWindows(windowSize), timestampedData, 15, 15, 15)
```
Doesn't the "Always" trigger happen on every element? Should we be actually expecting the sums, or the individual elements?  (I don't know what's correct here, outside of the default trigger.)

riteshghorse replied:
Yes. I think there is something missing in the trigger implementation. Will be working on it next.

lostluck commented on sdks/go/pkg/beam/core/graph/coder/panes.go:
```diff
@@ -0,0 +1,143 @@
+)
+
+const (
+	FIRST     int = 0
+	ONE_INDEX int = 1
+	TWO_INDEX int = 2
```
In Go, constants, exported or not should use MixedCaps if they're Exported or mixedCaps if they're unexported never ALL_CAPS.

lostluck commented on sdks/go/pkg/beam/core/graph/coder/panes.go:
```diff
@@ -0,0 +1,143 @@
+	}
+}
+
+// EncodePane encodes a single byte.
```
```suggestion
// EncodePane encodes a typex.PaneInfo.
```

lostluck commented on sdks/go/pkg/beam/core/graph/coder/panes.go:
```diff
@@ -0,0 +1,143 @@
+	if b&0x02 == 2 {
+		pn.IsLast = true
+	}
+	switch int64((b >> 2) & 0x03) {
+	case 0:
+		pn.Timing = typex.EARLY
+	case 1:
+		pn.Timing = typex.ON_TIME
+	case 2:
+		pn.Timing = typex.LATE
+	case 3:
+		pn.Timing = typex.UNKNOWN
+	}
```
If you define those arrival conditions (early late etc)  as based on integers or uints with iota, you can replace this whole switch with: `pn.Timing = typex.Timing((b >> 2) & 0x03)`

lostluck commented on sdks/go/pkg/beam/core/graph/coder/panes.go:
```diff
@@ -0,0 +1,143 @@
+		return typex.PaneInfo{}, err
+	}
+	pn := NewPane(data[0] & 0x0f)
+	switch encodingType(data[0]) {
```
Cleaning up this switch statement I'll leave as an excercise based on my other comments.

lostluck reviewed (changes requested):
I have many comments, but this is excellent progress. You've found many of the correct places to change without additional guidance.

We'll need maybe one or two more passes after these changes to sort out user side things and further initial testing, but this is great.

riteshghorse commented:
A lot of clean up is remaining in translate.go and possibly in strategy.go and trigger.go.
Added windowing strategy params, trigger params. Tests for Working Triggers.

riteshghorse commented:
PTAL. Basic triggers. Will add more tests in windowinto as they are similar to what we currently have.

lostluck commented on sdks/go/pkg/beam/core/runtime/graphx/translate.go:
```diff
@@ -981,22 +981,114 @@ func marshalWindowingStrategy(c *CoderMarshaller, w *window.WindowingStrategy) (
+		ClosingBehavior:  pipepb.ClosingBehavior_EMIT_IF_NONEMPTY,
+		AllowedLateness:  10,
+		OnTimeBehavior:   pipepb.OnTimeBehavior_FIRE_ALWAYS,
+		EnvironmentId:    "",
```
We can also keep this unset.

lostluck commented on sdks/go/pkg/beam/core/runtime/graphx/translate.go:
```diff
@@ -981,22 +981,114 @@ func marshalWindowingStrategy(c *CoderMarshaller, w *window.WindowingStrategy) (
+		AccumulationMode: accMode,
+		OutputTime:       pipepb.OutputTime_EARLIEST_IN_PANE,
+		ClosingBehavior:  pipepb.ClosingBehavior_EMIT_IF_NONEMPTY,
+		AllowedLateness:  10,
```
This should default to 0, unless we're pulling it from the passed in strategy.

lostluck commented on sdks/go/pkg/beam/core/runtime/graphx/translate.go:
```diff
@@ -981,22 +981,114 @@ func marshalWindowingStrategy(c *CoderMarshaller, w *window.WindowingStrategy) (
-		Trigger: &pipepb.Trigger{
+		Trigger:          trigger,
+		AccumulationMode: accMode,
+		OutputTime:       pipepb.OutputTime_EARLIEST_IN_PANE,
```
This changed from OutputTime_END_OF_WINDOW to this. We probably don't want that right now.

lostluck commented on sdks/go/pkg/beam/core/runtime/graphx/translate.go:
```diff
@@ -981,22 +981,114 @@ func marshalWindowingStrategy(c *CoderMarshaller, w *window.WindowingStrategy) (
 	} else {
 		mergeStat = pipepb.MergeStatus_NON_MERGING
 	}
+	trigger := makeTrigger(w.Trigger)
+	accMode := makeAccumulationMode(w.AccumulationMode)
```
We can inline these calls to the proto field assignments instead of assigning to variables first.

lostluck commented on sdks/go/pkg/beam/core/typex/special.go:
```diff
@@ -64,6 +65,19 @@ type Window interface {
+)
+
+type PaneInfo struct {
+	Timing                     int
```
As much as it's probably extra, we don't want this to be an int in type, as then users might try to manipulate it as an int.
Instead add a new type `type PaneTiming byte`  (since we really won't need more than that presently, or likely, ever) then the above enum declaration can be

```
const (
	PaneEarly   PaneTiming = 0 
	PaneOnTime  PaneTiming = 1 
	PaneLate    PaneTiming = 2 
	PaneUnknown PaneTiming = 3 
)
```
(Since we have specific definitions, we can't lean on an iota, even though it comes out the same. We want them to match their defined meaning in the coder definition).

The net result is that in order to use it in math/etc, a user would need to explicitly type cast to a numeric value. This avoids being able to misuse the enum type as a value rather than symbolically which is all it needs to have. 

This is necessary since the PaneInfo will eventually exposed to user DoFns, and it needs to be hard to misuse, which we're doing with the type system.

lostluck commented on sdks/go/test/integration/primitives/windowinto.go:
```diff
@@ -91,3 +92,60 @@ func WindowSums_GBK(s beam.Scope) {
+	col := teststream.Create(s, con)
+	windowSize := 10 * time.Second
+	validate(s.Scope("Fixed"), window.NewFixedWindows(windowSize), col, window.Trigger{Kind: window.DefaultTrigger}, window.Accumulating, 6.0, 9.0)
+
```
rm spare blank line.

lostluck commented on sdks/go/test/integration/primitives/windowinto_test.go:
```diff
@@ -36,3 +36,24 @@ func TestWindowSums_GBK(t *testing.T) {
 }
+
+func TestTriggerDefault(t *testing.T) {
+	integration.CheckFilters(t)
```
This has been mentioned elsewhere, since we're using TestStream here, we need to filter these tests out from everything but the Flink runner. Fortunately, we can use regex and just cover the prefix "TestTrigger"

lostluck reviewed (changes requested):
I think this is the home stretch.

lostluck commented on sdks/go/test/integration/integration.go:
```diff
@@ -69,6 +69,8 @@ var directFilters = []string{
 	// The portable runner does not support the TestStream primitive
 	"TestTestStream.*",
+	// The trigger tests uses TestStream
+	"TestTrigger.*",
```
You'll also need to add this to the samza, spark, direct, and dataflow tests. Basically anywhere TestStream is being filteed out right now.

riteshghorse commented on sdks/go/test/integration/primitives/windowinto.go:
```diff
@@ -91,3 +92,60 @@ func WindowSums_GBK(s beam.Scope) {
+}
+
+// TriggerAlways tests the Always trigger, it is expected to receive every input value as the output.
+// It also return an extra empty pane. Not sure why it is so. It is only in the case of this trigger
```
Got it fixed. It was because OnTimeBehavior was set to fire always. Now changed it to fire if non-empty. Will push that changes soon.

riteshghorse commented on sdks/go/test/integration/primitives/windowinto.go:
```diff
@@ -91,3 +92,60 @@ func WindowSums_GBK(s beam.Scope) {
+	windowed := beam.WindowInto(s, window.NewGlobalWindows(), col, beam.WindowTrigger{Name: tr}, beam.AccumulationMode{Mode: window.Discarding})
+	sums := stats.Sum(s, windowed)
+	sums = beam.WindowInto(s, window.NewGlobalWindows(), sums)
+	passert.Count(s, sums, "total collections", 1)
```
yes, you're right

riteshghorse commented:
Resolved the comments and added a test for Repeat Trigger.
PTAL.

riteshghorse commented:
PTAL. AfterAll and AfterAny Triggers are throwing Illegal Argument Exception. Working on it.

lostluck commented on sdks/go/pkg/beam/core/runtime/exec/datasource.go:
```diff
@@ -144,8 +144,8 @@ func (n *DataSource) Process(ctx context.Context) error {
 		}
-		// TODO(lostluck) 2020/02/22: Should we include window headers or just count the element sizes?
-		ws, t, err := DecodeWindowedValueHeader(wc, r)
+    // TODO(lostluck) 2020/02/22: Should we include window headers or just count the element sizes?
```
Probably need to run gofmt again.

lostluck commented on sdks/go/pkg/beam/core/runtime/graphx/translate.go:
```diff
@@ -981,22 +981,135 @@ func marshalWindowingStrategy(c *CoderMarshaller, w *window.WindowingStrategy) (
+			},
+		}
+	case window.AfterProcessingTimeTrigger:
+		// TODO: Right now would work only for single delay value.
```
Prefer not to leave unreferenced TODOs. Add the JIRA (BEAM-3304 ) for these or file a new one for these specific TODOs and list them as subtasks to BEAM-3304

lostluck commented on sdks/go/pkg/beam/windowing.go:
```diff
@@ -21,21 +21,48 @@ import (
+		case AccumulationMode:
+			ws.AccumulationMode = opt.Mode
+		default:
+			panic("Invalid option for Windowing Strategy")
```
Since WindowIntoOptions are implementers of interfaces, we can improve this by actually communicating what's going wrong. That is something like `panic(fmt.Sprintf("Unknown WindowingInto option type: %T: %v", opt, opt))` which communicates the type, and if it's printable somehow so it's easier to find what needs working on.

Since an error like this is likely an SDK dev error, a panic is fine, as there's nothing a user can do about it programmatically.

lostluck reviewed (approved):
This is good for now once the last small nits are worked out. Then we can work out the leftover details in smaller PRs.

riteshghorse commented:
Run Go PostCommit

lostluck commented:
Run Go Flink ValidatesRunner

lostluck commented:
Run Go PostCommit

lostluck commented:
It's not clear what's causing the failure with flink in this case. It's almost certainly the same TestStream issues. Lets disable the tests on flink entirely for now (like they are for the other runners) and we'll investigate it properly from there.

lostluck commented on sdks/go/test/integration/primitives/windowinto.go:
```diff
@@ -19,12 +19,12 @@ import (
 
-	"github.com/apache/beam/sdks/go/pkg/beam"
 	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/mtime"
 	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/window"
```
Looks like these have been missed.

riteshghorse replied:
Done.

lostluck commented on sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go:
```diff
@@ -22,18 +22,18 @@ import (
 	"path"
 
-	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/coder"
 	"github.com/apache/beam/sdks/go/pkg/beam/core/graph/mtime"
```
Looks like these have been missed.

riteshghorse commented:
Run Go Flink ValidatesRunner

riteshghorse commented:
Run Go PostCommit
